feat: create account from Package

lima-limon-inc opened the pull request:
This PR replaces the `--component-template` flag with a `--package` flag which allows for Account creation from a Miden `.masp` `Package` instead of an `Account Component Template` `.mct` file.

Additionally, it expands upon the error diagnostics present in the `miden-client init`; which was marked as a TODO item in this [comment](https://github.com/0xMiden/miden-client/pull/1313/files#diff-2e64311e3096e07a848fa04d15248cd63ead85a412613a78d3023d54d41f2aa9L123-L124).

igamigo commented on bin/miden-cli/src/commands/new_account.rs:
```diff
@@ -19,6 +20,7 @@ use miden_objects::account::{
     InitStorageData,
     StorageValueName,
 };
+use miden_objects::vm::Package;
```
This should be re-exported form `miden-client` instead

igamigo commented:
Code looks good but as we discussed offline I wonder if we should revisit the approach. Specifically, not sure if it makes sense to me to give the option to take both component templates as inputs as well as `Package` now. Conceptually, for instantiating accounts in the context of the CLI I think we have at least a couple of options:

- Remove the concept of component template (or `.mct` files) and make everything `Package`-based
- Remove the concept of `Package`-based templates (basically what was added here) and make everything component template-based. Here we could have the option to "import" a `Package` which then gets converted to an account component template and only then the user could instantiate accounts with the component template

Because the `Package` could be more than a component template (AFAIK), I think the second option makes the most sense.

igamigo commented:
This should also close #1132 (or part of it)

bobbinth commented:
> Right now we pre-build some (all? not entirely sure) of the well-known account components as component templates so that they can be used from the CLI. These is done through the `.mct` files which are basically serialized `AccountComponentTemplate`s (so, the component code and the storage layout description) which are placed in a specific, configurable path. If we remove the concept of these files, then we'd need to provide them as `Package`s so that they can be used in the way that is implemented in this PR (or look into providing them some other way as well).

Is the difficulty here mainly in having to "manually" build the packages? Other than that, things should be relatively unchanged - i.e., we could still have packages for the "well-known" components and place them into a specific config directory etc.

bobbinth commented:
The way I think about it is that `AccountComponentTemplate` is a specialization of a `Package`. Basically, every `AccountComponentTemplate` is a `Package` but not every `Package` is an `AccountComponentTemplate`.

But I agree, we can probably get rid of `AccountComponentTemplate` because the main important part there is `AccountComponentMetadata` and so, if we keep that, we can always read/validate a given package for purposes of account component creation.

So, I largely agree with the changes in @mmagician's branch:

We keep `AccountComponent::from_package_with_init_data()` (and probably just rename it to `AccountComponent::from_package()`). We might still want to keep `AccountComponent::new()` for tests and other ad-hock purposes where creating a package is a bit cumbersome. But the main way to instantiate components would be from packages.

mmagician commented:
> I think I'm unaware of this - how is this currently involved in FPI?

Oh perhaps I was not precise: the way we **currently** instantiate components [in our FPI tests](https://github.com/0xMiden/miden-client/blob/6dcae62c64f5aa21ce30ead8ed86cefba4d48d6a/bin/integration-tests/src/tests/fpi.rs#L359) is via `compile`.

igamigo commented on bin/miden-cli/src/commands/new_account.rs:
```diff
@@ -150,9 +153,16 @@ pub struct NewAccountCmd {
+                This field is only kept to inform users about said change."
+    )]
+    #[arg(short, long, hide(true))]
     pub component_templates: Vec<PathBuf>,
```
Should we remove this for now? No need to mark deprecations while we are developing unstable versions.

lima-limon-inc replied:
Noted! Removed in this commit: https://github.com/0xMiden/miden-client/pull/1313/commits/4e8764e6d02ce4a1d22dcfe8576908f0118fba55

igamigo commented on bin/miden-cli/src/commands/new_account.rs:
```diff
@@ -172,12 +182,26 @@ impl NewAccountCmd {
+",
+                client_binary_name().display()
+            )));
+        }
```
Same here

lima-limon-inc replied:
Same as above (https://github.com/0xMiden/miden-client/pull/1313/commits/4e8764e6d02ce4a1d22dcfe8576908f0118fba55)

igamigo commented on bin/miden-cli/src/config.rs:
```diff
@@ -28,8 +28,8 @@ pub struct CliConfig {
     pub remote_prover_endpoint: Option<CliEndpoint>,
-    /// Path to the directory from where account component template files will be loaded.
-    pub component_template_directory: PathBuf,
+    /// Path to the directory from where [[`miden_core::vm::Package`]]s will be loaded.
```
I don't think we should use Rustdoc style links in the commands as they are mostly useful with regards to the CLI usage itself (and they would not render in a very useful way). If we wanted to add more context we can add a link to package docs (though I'm not sure we have user-facing docs for that)

lima-limon-inc replied:
Got it, changed in this commit: https://github.com/0xMiden/miden-client/pull/1313/commits/440b38671203801a08d20fa77871fb577d38b771

igamigo commented on docs/src/cli-config.md:
```diff
@@ -87,29 +87,10 @@ Notice how the amount specified when using the token symbol takes into account t
-### Component template directory
-The `component_template_directory` field is used to configure the path to the directory where the account component templates are stored. The default value is `./templates`.
+### Package directory
+The `package_directory` field is used to configure the path to the directory where the account component are stored in package (`.masp`) form. The default value is `./packages`.
```
Would be great to expand on what a `Package` is, but not sure this is the right place. Maybe we can link somewhere? Only specific docs I found were [these](https://0xmiden.github.io/compiler/appendix/known-limitations.html#packaging)

igamigo reviewed (approved):
Looks great! Thanks. Left some minor comments.

lima-limon-inc commented:
> Looks great! Thanks. Left some minor comments.

Thanks 😊 ! I'll get to work on addressing your comments + updating this PR to the newest version of `miden-base` since https://github.com/0xMiden/miden-client/pull/1382 got merged. I'll mark it as a draft until then. 

If I'm not mistaken, the main gist of this PR should remain the same.

lima-limon-inc commented on Cargo.toml:
```diff
@@ -29,23 +29,23 @@ opt-level = 1
+miden-lib     = { branch = "fabrizioorsi/re-export-Section", default-features = false, git = "https://github.com/lambdaclass/miden-base" }
+miden-objects = { branch = "fabrizioorsi/re-export-Section", default-features = false, git = "https://github.com/lambdaclass/miden-base" }
+miden-testing = { branch = "fabrizioorsi/re-export-Section", default-features = false, git = "https://github.com/lambdaclass/miden-base" }
+miden-tx      = { branch = "fabrizioorsi/re-export-Section", default-features = false, git = "https://github.com/lambdaclass/miden-base" }
```
I created this PR https://github.com/0xMiden/miden-base/pull/2015 in miden-base to re-export the required structures.

mmagician commented on crates/rust-client/src/account/mod.rs:
```diff
@@ -75,6 +75,7 @@ use crate::sync::NoteTagRecord;
 
 pub mod component {
     pub const COMPONENT_TEMPLATE_EXTENSION: &str = "mct";
```
```suggestion
```

mmagician replied:
this can be removed now right?

mmagician reviewed (approved):
LGTM ✅
